### PR TITLE
Raycaster: Fixed Points with material.sizeAttenuation set to false

### PR DESCRIPTION
**Pull request by anhr**
I think Raycaster works incorrectly for points with material.sizeAttenuation = false. Area of the intersection of mouse with point is very big if distance from point to camera is little.
I moving a point along ray to distance from camera to new Vector3() if object.material.size Attenuation is false for resolving of issye.

### CODE REVIEW

**Comment by WestLangley** [comment]
Intersection is based on the distance from the  _center_ of the point to the ray. The size and shape of the rendered point is irrelevant.

**Comment by anhr** [comment]
> Intersection is based on the distance from the _center_ of the point to the ray. The size and shape of the rendered point is irrelevant.

I do not agree with you. Please see my [example](https://raw.githack.com/anhr/commonNodeJS/master/Examples/webgl_raycaster.html) . You can see two points in my example.

Mouse cusor is changing to 'pointer' if you move mouse over points. Also you can see an alert if you click a point.

Left point with material.sizeAttenuation set to true. Left point is visible as big square. It point have intersection area inside of the point. It looks good.

Right point with material.sizeAttenuation set to false. Right point is visible as small square. It point have vary big intersection area outside of the point similar as left point. It looks very bad.

I have resolved this issue by editing of the Points.testPoint function.

**Comment by WestLangley** [comment]
>Intersection is based on the distance from the center of the point to the ray. The size and shape of the rendered point is irrelevant.
>>I do not agree with you

As currently implemented, `Points` detection occurs when the distance to the ray is less than the specified `threshold`. The size of the rendered point is irrelevant.

Maybe we could consider making the threshold vary with distance...

If you need "pixel-perfect" accuracy, then try a `Sprite`, instead. See https://threejs.org/examples/webgl_raycast_sprite.html.

**Comment by anhr** [comment]
> > Intersection is based on the distance from the center of the point to the ray. The size and shape of the rendered point is irrelevant.
> > > I do not agree with you
> 
> As currently implemented, `Points` detection occurs when the distance to the ray is less than the specified `threshold`. The size of the rendered point is irrelevant.
> 
> Maybe we could consider making the threshold vary with distance...
> 
> If you need "pixel-perfect" accuracy, then try a `Sprite`, instead. See https://threejs.org/examples/webgl_raycast_sprite.html.

Yes of course the size of the rendered point is irrelevant.

But threshold from ray to point appears larger near the camera and small far from the camera. It is good for points with material.sizeAttenuation set to true.

I can making the threshold vary with distance to camera as you said. But I have desided it is very difficult.

Instead I moving all points along ray to distance from camera to new Vector3( 0, 0, 0 ) position if  material.sizeAttenuation is false. In another words all points have same distance to camera if  material.sizeAttenuation is false. As result the threshold from ray to point seems to be the same for all points.

**Comment by WestLangley** [comment]
@anhr Out of curiosity, is this a serious problem for your use-case -- or are you reporting it because you just discovered it?

If it is a serious problem, can you use sprites instead?

**Comment by anhr** [comment]
> @anhr Out of curiosity, is this a serious problem for your use-case -- or are you reporting it because you just discovered it?
> 
> If it is a serious problem, can you use sprites instead?

I have detected problem in the Points very long time ago, resolved it and have desided to make a pull request. Do you agree it is problem of the Points? I think, good idea to fix this bug. I want to use Points (not Sprites) in my project and do not want to see this bug in the threejs.

**Comment by WestLangley** [comment]
I agree that the current implementation has limitations.

Any proposed changes must be an improvement for orthographic cameras, and for perspective cameras with `sizeAttenuation` set to `true` or `false`.

/ping @Mugen87 for your opinion on this issue.

File: src/objects/Points.js [hunk 1]
```diff
@@ -160,6 +160,9 @@ Points.prototype = Object.assign( Object.create( Object3D.prototype ), {
 
 function testPoint( point, index, localThresholdSq, matrixWorld, raycaster, intersects, object ) {
 
+	if ( ! object.material.sizeAttenuation )
+		new Ray( _ray.origin ).lookAt( point ).at( new Vector3().distanceTo( _ray.origin ), point );
```
Comment (WestLangley) [inline src/objects/Points.js:164]: Tip: Avoid instantiating new objects -- particularly in methods that are called in tight loops. Create single instances and reuse them.

Comment (anhr) [reply src/objects/Points.js]: Hope I understand you.
See my new version [testSizeAttenuation](https://github.com/mrdoob/three.js/pull/20144/commits/eecfc03e68845a695c5e88f4c50a92b216b02f1c)

**Comment by anhr** [comment]
Now you can set the distance from point to camera. Example:
```
const points = new THREE.Points( new THREE.BufferGeometry().setFromPoints( [
	new THREE.Vector3( 0.3, 0, 4.5 ),
] ),
	new THREE.PointsMaterial( {

		color: 0xffffff,
		size: 5,//0.05,
		sizeAttenuation: false,
		alphaTest: 0.5,

	} ) );
points.testSizeAttenuation.distanceToCamera = 2;
```
The `points.testSizeAttenuation.distanceToCamera` allow you to set the area of intersetion of the mouse pointer with point.
A small distance increases the area of intersection.
Apply only to Points with material.sizeAttenuation set to false.

**Comment by anhr** [comment]
> I agree that the current implementation has limitations.
> 
> Any proposed changes must be an improvement for orthographic cameras, and for perspective cameras with `sizeAttenuation` set to `true` or `false`.
> 
> /ping @Mugen87 for your opinion on this issue.

I have sucessfully tested for orthographic camera. See [webgl_raycaster_orthographic_camera.html](https://github.com/mrdoob/three.js/pull/20144/commits/e8ed516eed3c24535e7ef1855f4b4219d8e8a82a)

File: src/objects/Points.js [hunk 2]
```diff
@@ -188,4 +190,22 @@ function testPoint( point, index, localThresholdSq, matrixWorld, raycaster, inte
...
 
+function testSizeAttenuation() {
+
+	const ray = new Ray( _ray.origin );
+	this.distanceToCamera;
+	this.testPoint = function ( point, object ) {
+
+		if ( object.material.sizeAttenuation )
+			return;
+		if ( ! this.distanceToCamera )
```
Comment (mrdoob) [inline src/objects/Points.js:201]: Is this what we're checking?

```js
if ( this.distanceToCamera === undefined )
```

**Comment by mrdoob** [comment]
The current approach seems hacky...

@WestLangley if you were to fix this, how would you do it?

**Comment by WestLangley** [comment]
@mrdoob I would ask @06wj if he was interested in looking at this, as he did with https://threejs.org/examples/webgl_raycast_sprite.html.

**Comment by 06wj** [comment]
I am interested in this, and I will fix it in a sprite-like way, which will be more precise.

**Comment by WestLangley** [comment]
@06wj Just remember that points can have textures with alpha, so I would not worry too much about being pixel-perfect. Just get the scale about right, if you can.

**Comment by 06wj** [comment]
Because gl_PointSize is in the screen coordinates, if we want to calculate accurately, we need to get the screen information.
I don’t know how to get the screen information in raycaster.

**Comment by mrdoob** [comment]
@06wj

Shouldn't it be possible to add `material.sizeAttenuation = true` support by using the camera's frustum?

We have access to the camera:

https://github.com/mrdoob/three.js/blob/fa7b498aca396f8d23d8264f886bdbceff4ccf7f/src/core/Raycaster.js#L73-L93

**Comment by Mugen87** [comment]
> The current approach seems hacky...

Closing then. A different approach is required.

**Pull request closed without merging**